### PR TITLE
FS - leadgen assistant handling

### PR DESCRIPTION
**Pull request by fszwedo**
Functionality to expose API endpoint to consume leadgens from assistant.

Can be used to handle support reports from the assistant as https://orca-preview.zoovu.com/?c=bwhkcHf7&locale=en-IE&type=WEB_CLICK

When reviewing please ignore ticketCreationServiceTestSamples.ts - its just a file containing snapshots of leadgen requests and expected responses - this was the simplest way I came up with to test the ticket creation service (but probably will be a pain to maintain in the future).

### CODE REVIEW

File: src/controllers/ticketController.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        let ticket;
+        //use appropriate service method for the selected path
+        try {
+            if (ticketCategory === 'Account access') ticket = await this.service.createAccountAccessRequest(req.body);
+            else if (ticketCategory === 'Admintool access') ticket = await this.service.createAdminAccessRequest(req.body);
+            else if (ticketCategory === 'Account creation') ticket = await this.service.createAccountCreationRequest(req.body);
+            else if (ticketCategory === 'Platform issue') ticket = await this.service.createProblemReport(req.body);
+            else if (ticketCategory === 'FTP') ticket = await this.service.createFTPRequest(req.body);
+            else if (ticketCategory === 'Reporting') ticket = await this.service.createReportingRequest(req.body);
+            else ticket = await this.service.createGeneralTicket(req.body);
```
[the remark this answers is not in the document]
Comment (fszwedo) [reply src/controllers/ticketController.ts:24]: applied the proposed change, thanks!

File: src/services/zendesk/ticketCreationService.ts [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+            let ticketComment = {
+                comment: {
+                    html_body: `WARNING - it was not possible to extract the email of the requester automatically. Phil was set as requester, please adjust it accordingly!`,
+                    "public": false
+                }
+            }
+            await sendZendeskTicketUpdateRequest(ticketComment, createdTicket.ticket.id)
+        }
+
+        return createdTicket
```
[the remark this answers is not in the document]
Comment (fszwedo) [reply src/services/zendesk/ticketCreationService.ts:86]: love it :) applied the suggested change @rajashree27

**Review by rajashree27 — APPROVED** [review]

File: src/services/zendesk/ticketCreationService.ts [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+    }
+
+    createAccountAccessRequest = async (ticketData: leadgenFormContent) => {
+        const accountLink = ticketData.submittedFormData.find(el => el.questionText === "Account link").answers[0];
+        const requesterEmail = ticketData.submittedFormData.find(el => el.questionText.toLowerCase() === "please provide your email").answers[0];
+        const userToBeAssigned = ticketData.submittedFormData.find(el => el.questionText.includes("should get access")).answers[0];
+        const approverEmail = ticketData.submittedFormData.find(el => el.questionText.includes('line manager')).answers[0];
+        const platform = ticketData.questionsFlow.find(el => el.questionText.includes('which environment')).answers[0];
```
[the remark this answers is not in the document]
Comment (fszwedo) [reply src/services/zendesk/ticketCreationService.ts]: added!

File: src/models/ticketModel.ts [hunk 4]
```diff
@@ -6,5 +6,28 @@
...
+      }[]
+      type?: string | null
+      requester:{
+        email: string,
+        name?: string
+      }
+      tags? : string[]
+      email_ccs?: {
+        user_email: string,
+        user_name: string
```
[the remark this answers is not in the document]
Comment (fszwedo) [reply src/models/ticketModel.ts:27]: honestly no idea, also i dont know how does this work without the comma :D I added it

File: index.ts [hunk 5]
```diff
@@ -42,44 +43,52 @@ const shiftChangeRepository = new ShiftChangeRepository(shiftChangeRequestModel)
...
 
+
 const logger = new LoggerService(new LoggerRepository(logModel));
 
 const mongooseConnection = async () => {
     await mongoose.connect(`mongodb+srv://${process.env.MONGOLOGIN}:${process.env.MONGOPW}@cluster0.mgkhb.mongodb.net/myFirstDatabase?retryWrites=true&w=majority`)
-    .then(() => console.log('Connected to MongoDB...'))
-    .catch(error => console.error('Could not connect to MongoDB!', error))
+        .then(() => console.log('Connected to MongoDB...'))
+        .catch(error => console.error('Could not connect to MongoDB!', error))
```
[the remark this answers is not in the document]
Comment (fszwedo) [reply index.ts:78]: I think its added by my formatter automatically :)

File: src/controllers/ticketController.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    constructor(service: TicketService) {
+        this.service = service;
+    }
+
+    postTicket = async (
+        req: express.Request,
+        res: express.Response,
+    ) => {
+        const ticketCategory = req.body.questionsFlow[0].answers[0];
+        let ticket;
```
[the remark this answers is not in the document]
Comment (fszwedo) [reply src/controllers/ticketController.ts:15]: I added string - technically I could use an enum here but as the available options are listed in the switch statement later in the file I don't see the benefit in that (and I don't plan on using category anywhere else)

File: src/models/ticketModel.ts [hunk 7]
```diff
@@ -6,5 +6,28 @@
     level:'L1' | 'L2' | 'other' | 'any'
```
[the remark this answers is not in the document]
Comment (fszwedo) [reply src/models/ticketModel.ts:6]: I guess we can keep it - I'd like to rework how the categorization works to implement the logic that distributes the tickets according to the expertise, so let's touch this area then

File: utilities/getZendeskTicketFields.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    const fields = await makeZendeskRequest('/api/v2/ticket_fields.json', 'GET')
+    console.log(fields)
+}
+
+const getZendeskTicketField = async () => {
+    const field = await makeZendeskRequest('/api/v2/ticket_fields/360016765134', 'GET')
+    console.log(field)    
+    console.log(field.ticket_field.custom_field_options)
+}
+
```
[the remark this answers is not in the document]
Comment (fszwedo) [reply utilities/getZendeskTicketFields.ts:15]: this is a utility function that has to return the ticket fields if I'll ever need to update them in constants - console.log is the key element here so let's keep it :)

**Pull request merged**